lib/ansible/modules/clustering/kubevirt/k8s_service.py

```diff
@@ -0,0 +1,208 @@
+from ansible.module_utils.k8s.raw import KubernetesRawModule
+
+from openshift import watch
+from openshift.dynamic.client import ResourceInstance
+from openshift.helper.exceptions import KubernetesException
+
+
+SERVICE_ARG_SPEC = {
+    'merge_type': {'type': 'list', 'choices': ['json', 'merge', 'strategic-merge']},
+    'state': {
+        'type': 'str',
+        'choices': [
+            'present', 'absent'
+        ],
+        'default': 'present'
+    },
+    'selector': {'type': 'dict', 'default': {}},
+    'type': {
+        'type': 'str',
+        'choices': [
+            'NodePort', 'ClusterIP', 'LoadBalancer'
+        ],
+        'default': 'ClusterIP'
+    },
+    'ports': {'type': 'list', 'default': []},
+    'labels': {'type': 'dict', 'default': {}},
+    'spec': {'type': 'dict', 'default': {}},
+}
+
+API_VERSION = 'v1'
+
+
+def virtdict():
+    return defaultdict(virtdict)
+
+
+class KubernetesService(KubernetesRawModule):
+    def __init__(self, *args, **kwargs):
+        super(KubernetesService, self).__init__(*args, **kwargs)
+
+    @staticmethod
+    def merge_dicts(x, y):
+        for k in set(x.keys()).union(y.keys()):
+            if k in x and k in y:
+                if isinstance(x[k], dict) and isinstance(y[k], dict):
+                    yield (k, dict(KubernetesService.merge_dicts(x[k], y[k])))
+                else:
+                    yield (k, y[k])
+            elif k in x:
+                yield (k, x[k])
+            else:
+                yield (k, y[k])
+
+    @property
+    def argspec(self):
+        """ argspec property builder """
+        argument_spec = copy.deepcopy(COMMON_ARG_SPEC)
+        argument_spec.update(copy.deepcopy(AUTH_ARG_SPEC))
+        argument_spec.update(SERVICE_ARG_SPEC)
+        return argument_spec
+
+    def execute_module(self):
+        """ Module execution """
+        self.client = self.get_api_client()
+
+        selector = self.params.get('selector')
+        service_type = self.params.get('type')
+        ports = self.params.get('ports')
+        labels = self.params.get('labels')
+        spec = self.params.get('spec')
+
+        definition = virtdict()
+
+        def_spec = definition['spec']
+        def_spec['type'] = service_type
+        def_spec['ports'] = ports
+        def_spec['selector'] = selector
+
+        def_meta = definition['metadata']
+        def_meta['labels'] = labels
+
+        # 'spec' takes precedence
+        definition = dict(self.merge_dicts(definition, spec))
+
+        # Perform present/absent action:
+        definition = dict(self.merge_dicts(self.resource_definitions[0], definition))
+
+        resource = self.find_resource('Service', self.api_version, fail=True)
+        definition = self.set_defaults(resource, definition)
+        result = self.perform_action(resource, definition)
+        changed = result['changed']
+
+        self.exit_json(**{
+            'changed': changed,
+            'service': result,
```

Comment:
Not sure if it's part of the result, but I think we should definitelly return user-friendly IP address, based on service type

Reply:
Good thing I looked at this. We need to change the way we run `exit_json()`. Currently we produce `result.result.*`. Now I've changed it to just `result.*`.

Back to your question. Currently one gets that info with `result.spec.clusterIP`. I don't really see the point of adding a second name for that value. Let's stick with what k8s does if it looks clear enough.

Reply:
Ok,  I am fine if it's nested.